### ISSUE TITLE
How to dispatch in event handler?

### ISSUE THREAD

**Issue by pke** [yelouafi/redux-saga#421]
I have this function that I call inside one of my sagas

```es6
function getDeviceTokenAsync() {
  return new Promise((s,e) => {
    if (window.PushNotification) {
      let push = window.PushNotification.init({
        android: {
          senderID: GCM_SENDER_ID,
          icon: "ic_notification",
          iconColor: PRIMARY_COLOR,
        },
        ios: {
          alert: true,
          badge: false,
          sound: false,
        },
        windows: {
        },
      })
      push.on("registration", s)
      push.on("notification", (data) => {
        console.info("notification: " + JSON.stringify(data))
        //put(actions.refreshPendingTransactions())
      })
      push.on("error", e)
    } else {
      e(new Error("No push notification available"))
    }
  })
}
```

I call it with
`yield call(getDeviceTokenAsync)`

The question is, how do I dispatch an action to the store whenever the `onnotification` handler is called by the push service?
Should I hand in a function to `getDeviceTokenAsync` and that function yields a `put` effect then?

**Comment by slorber** [comment]
Hi,

I'm using the same Cordova push plugin :)

Here it's a bit particular because you want this function to both return a Apns/GCM token, and a stream of future notifications. It's not the responsability of this method to be coupled with redux and dispatch actions.

Instead of returning just the token in the promise, why not also returning the push object, so that the caller can decide to subscribe to notifications? Alternatively you can probably create a channel and return it as well. The thing is that you probably don't want to return the channel until the registration is successful so you can put the channel inside the promise.

Honnetly, not sure it's important to return the stream/channel after the token. If the channel is available at the beginning but does not emit anything it's not a big deal imho. Also note that runnng your method multiple times will lead to multiple push plugin initializations. You might want to memoize the returned promise or ensure it only run once in your code.

**Comment by pke** [comment]
If returning the push object, can we call put effects in the event handler we attach to it?

**Comment by slorber** [comment]
hmmmm no I think in any case you need a channel to handle that properly :)

**Comment by pke** [comment]
And a channel is a redux-saga concept?

**Comment by slorber** [comment]
yes look at https://github.com/yelouafi/redux-saga/blob/master/docs/advanced/Channels.md#using-the-eventchannel-factory-to-connect-to-external-events

**Comment by pke** [comment]
Ok I found that. Which channel to use? Do you have a small example for my usecase, since you also use the push plugin maybe you have something ready?

**Comment by slorber** [comment]
There are some examples available where people tried to subscribe to websocket events from sagas, which is a similar usecase.

For example: http://stackoverflow.com/questions/35277682/how-to-tie-emitted-events-events-into-redux-saga

I have never used such channel yet however, and I'm not sure it's really worth it unless you really have to handle complex coordination.

Here's what I've done in my app: https://gist.github.com/slorber/88cc95d16647fd73fa071f497fd1c6f9

In my app, the push plugin simply triggers dispatches and could be simplied as:

```javascript
push.on('notification', function(data) {
        store.dispatch(AppEvents.mobilePushReceived(data));
});
```

I think this is largely enough for most push usecases.
Once the push event is dispatched, it's still possible to listen to that event in sagas and handle more complex workflows if needed.

**Comment by pke** [comment]
Where do you get the `store` from within the `push.onNotification`?

**Comment by slorber** [comment]
```javascript
function registerForPushes(store) {
  push.on('notification', function(data) {
          store.dispatch(AppEvents.mobilePushReceived(data));
  });
}

const store = createStore(reducer);
registerForPushes(store);
```

**Comment by pke** [comment]
aha! Well in my code I have no access to the store where I create the store. Hmm. I'd like to keep this all in the account saga that I have. There is no way the saga function has access to the store @yelouafi and could give the event handler the store?

**Comment by slorber** [comment]
Why would you not have access to the store? If you create a store, it is somewhere, and you can call a method passing that store as parameter.

**Comment by pke** [comment]
nah, my modules are much more separated. The store is created far far away from the sagas :) Anyway, I implemented a proper `PushNotificationSource` like in this SO posting http://stackoverflow.com/questions/34859932/can-i-use-redux-sagas-es6-generators-as-onmessage-listener-for-websockets-or-ev/34866840#34866840

```es6
function createPushSource(push, id) {
  let deferred

  push.on("notification", data => {
    console.info("notification: " + JSON.stringify(data))
    if (deferred) {
      deferred.resolve(data)
      deferred = null
    }
  })
  return {
    id,
    next() {
      if (!deferred) {
        deferred = {}
        deferred.promise = new Promise(resolve => deferred.resolve = resolve)
      }
      return deferred.promise
    },
  }
}

function* watchPushNotification(source) {
  let notification = yield call(source.next)
  while (notification) {
    yield put(actions.refreshPendingTransactions)
    notification = yield call(source.next)
  }
}

function initPushSupportAsync() {
  return new Promise((s,e) => {
    if (window.PushNotification) {
      let push = window.PushNotification.init({
        android: {
          senderID: GCM_SENDER_ID,
          icon: "ic_notification",
          iconColor: PRIMARY_COLOR,
        },
        ios: {
          alert: true,
          badge: false,
          sound: false,
        },
        windows: {
        },
      })
      push.on("registration", (data) => {
        console.info("PUSH registration: " + JSON.stringify(data))
        s(createPushSource(push, data.registrationId))
      })
      push.on("error", e)
    } else {
      e(new Error("No push notification available"))
    }
  })
}
```

**Comment by yelouafi** [comment]
@pke eventChannels are now part of redux-saga API

http://yelouafi.github.io/redux-saga/docs/advanced/Channels.html

In your example, it'd be something like

```js
function getDeviceTokenChannel() {
  if (window.PushNotification) {
    return eventChannel((emit) => {
      let push = window.PushNotification.init({...})
      push.on("notification", emit)
      push.on("error", e => emit({error: })
    })
 } else {
    throw new Error("No push notification available"))
 }
}

function* saga() {
  const pushChan = yield call(getDeviceTokenChannel)
  while(true) {
    const msg = yield take(pushChan)
    if(!msg.err) {
      yield put(actions.refreshPendingTransactions())
    }
  }
}
```

**Comment by pke** [comment]
@yelouafi thanks. The `window.PushNotification.init` has a success handler that is called with the push ID that I need to send to the server. So I'd have to somehow return from `getDeviceTokenChannel` only after the success callback has been called (like in my original code).  How would I incorporate this functionality?

**Comment by yelouafi** [comment]
you can return a promise of the channel from getDeviceTokenChannel

**Comment by yelouafi** [comment]
`eventChannel` returns a channel (cf API docs). A `call(getDeviceTokenChannel)` is a simple call to a function which returns either a value (of type channel) or a promise that will resolve to that value (promise of a channel). After that you can use the returned channel with `take` effects like shown above

```js
function getDeviceTokenChannel() {
  return new Promise((resolve, reject) => {
    if (window.PushNotification) {
      push.on('registration', (data) => resolve(eventChannel(...))
    }) else {
       reject("No push notification available"))
    }
  })
}
```

**Comment by pke** [comment]
as a follow up, I want to use channels to yield `document visibilitychange` events.

```es6
const eventChannel = (target, eventName) => emitter => {
  const listener = event => emitter(event)
  target.addEventListener(eventName, listener, false)
  return () => {
    target.removeEventListener(eventName, listener)
  }
}

function* refreshUser(access_token) {
  const visibilityChangeChannel = yield call(eventChannel, document, "visibilitychange")
  try {
    while (true) {
      yield call(apiUserRefresh, access_token)
      yield take([REFRESH_USER, visiblityChangeChannel])
    }
  } finally {
    if (yield cancelled()) {
      visibilityChangeChannel.close()
    }
  }
}
```

But the visibility event does not seem to be emitted.

**Comment by yelouafi** [comment]
You don't actually call `eventChannel` (and I mean the *real one* from redux-saga)

```es6
import { eventChannel } from 'redux-saga'

const createChannel = (target, eventName) => eventChannel/*here*/(emitter => {
  ...
})
```

Also taking from an array mixing channels and action types in not (yet!) supported. You should use race for this

```es6
function* refreshUser(access_token) {
  const visibilityChangeChannel = yield call(createChannel, document, "visibilitychange")
  try {
    while (true) {
      yield call(apiUserRefresh, access_token)
     /* here */
      yield race([take([REFRESH_USER), take(visiblityChangeChannel)])
   ...
  }
}
```

**Comment by pke** [comment]
@yelouafi Can eventChannel also throw exceptions instead of emitting a error object?

**Comment by pke** [comment]
@yelouafi One more thing to this concept you outlined here: https://github.com/yelouafi/redux-saga/issues/421#issuecomment-230475206

How would I close this channel (and terminate its saga that is using it) when the the user logs out?

**Comment by pke** [comment]
I tried this code now but it fails at `yield take(channel)` with

`Error: take(patternOrChannel): argument [object Object] is not valid channel or a valid pattern`

```es6
function* watchPushNotification(channel) {
  while (true) {
    const notification = yield take(channel)
    if (!notification.error) {
      yield put(actions.refreshPendingTransactions())
      yield put(actions.refreshInvoices())
    }
  }
}

const getDeviceTokenAsync = () => (
  new Promise((s,e) => {
    if (window.PushNotification && Product.googleSenderId) {
      const push = window.PushNotification.init({
        android: {
          senderID: Product.googleSenderId,
          icon: "ic_notification",
          iconColor: PRIMARY_COLOR,
        },
        ios: {
          alert: true,
          badge: false,
          sound: false,
        },
        windows: {
        },
      })
      push.on("registration", data => {
        __DEV__ && console.info("PUSH registration: " + JSON.stringify(data))
        const channel = eventChannel(emit => {
          push.on("notification", emit)
          push.on("error", error => emit({ error }))
          return () => {
            //push.unregister()
          }
        })
        debugger
        s({ channel, id: data.id })
      })
    } else {
      e(new Error("No push notification available"))
    }
  })
)

function* authorize(credentials) {
  try {
    const { email, password } = credentials
    const deviceInfo = getDeviceInfo()
    try {
      const { id, channel } = yield call(getDeviceTokenAsync)
      deviceInfo.device_token = id
      yield fork(watchPushNotification, channel)
    } catch (error) {
      // Silently ignore if no push service is avail
    }
    let { access_token, user } = yield call(api.loginAsync, email, password, deviceInfo)
    if (!user) {
      user = yield call(api.getUserAsync, access_token)
    }
    yield put(actions.loginSuccess(access_token, user))
  } catch (error) {
    console.error(error)
    yield put(actions.loginError(error))
  }
}
```

**Comment by Andarist** [comment]
hey, did you solve your issue? looking on the code you shouldnt receive this error `Error: take(patternOrChannel): argument [object Object] is not valid channel or a valid pattern`

**Comment by stoyandamov** [comment]
So if you're really really stuck:

```js
// globals.js
let cachedStore

export const setStore = store => cachedStore = store
export const getStore = () => cachedStore
```

Wherever you're creating the store:

```js
const store = createStore(
  combineReducers({...}),
  applyMiddleware(middleware, sagaMiddleware, logger)
)
setStore(store) // <--- this

```

Wherever you need dispatch:

```js
import {getStore} from '../globals'
const { dispatch } = getStore() // <--- this 
```

**Comment by pke** [comment]
@Andarist didn't solve it yet but had other things to do. @stoyandamov thats more like a hack :)

**Comment by Andarist** [comment]
Yeah, @stoyandamov's solution is definitely a hack and can be avoided pretty much always.

**Comment by malmator** [comment]
I couldnt get @stoyandamov solution to work but I have something very similar (see below). Nice and simple.
I want to be able to dispatch a redux action from within an event handler (onEnabledChange) that belongs to a plugin (cordova-background-geolocation-lt) that I start/stop in sagas. This should be straightforward. Perhaps there is a better solution. but not channels - too complicated for me
```
//  index.js
const store = createStore(rootReducer, composeEnhancers(applyMiddleware(...appMiddlewares)))
export const dispatch = store.dispatch

// saga.js
import { dispatch } from '../index.js'

// inside handler
 dispatch({ type: actions.MYACTION, afield: "some data", })

```
